Implement mapWithKeys, traverseWithKey, mapAccum, mapAccumWithKey and mapAccumRWithKey

YoEight opened the pull request:
There are also Foldable and Traverse instances for CritBit

Commit:
Implement map.

Commit:
Added a benchmark for map.

YoEight committed:
Implement mapWithKeys, traverseWithKey, mapAccum, mapAccumWithKey and mapAccumRWithKey
+ Foldable and Traverse Cribit instances

Fix build problems with GHC 7.4 (gh-1)

Thanks to Bob Ippolito for reporting this. Oops!

Relax version constraint on text

format

Mention what a crit-bit tree *is* (thanks, reddit!)

fix formatting + docs

bos commented:
Thanks for your patch. You'll notice that I have a bunch of comments, but they're all style and formatting and example comments that ought to be easy to fix.

Before you resubmit, please do a little investigating of whether adding `INLINABLE` pragmas makes a difference to benchmark performance. My expectation is that it will, and that most or all of the functions you have added will need them.

YoEight committed:
Implement mapWithKeys, traverseWithKey, mapAccum, mapAccumWithKey and mapAccumRWithKey

Fix build problems with GHC 7.4 (gh-1)

Thanks to Bob Ippolito for reporting this. Oops!

Relax version constraint on text

format

YoEight commented:
Thanks for your feedbacks. Please tell me if everything is ok since each time I'm doing a rebase something bad happens (it's the main reason why I'm not in favor of squashing commits I've already pushed on a remote repo)
